README.md
# Overview

[![Built with Grunt](https://cdn.gruntjs.com/builtwith.png)](http://gruntjs.com/)

Echo Recent Participants allows to display a facepile of recent participants.

# Installation instructions

0. Clone this repository.
1. Download and install NodeJS and npm from [http://nodejs.org](http://nodejs.org) (either from package or the source).
2. sudo npm install grunt-cli -g
3. Run the following command in the project directory: npm install
4. Now you can use 'grunt' command to build the project

# License

Copyright 2014 Echo

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this work except in compliance with the License. You may obtain a copy of the License in the LICENSE file, or at:

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


Changelog.md
# Echo RecentParticipants CHANGELOG:

##v1.0.6 - February 23, 2015

* App Title was changed from "RecentParticipants" to "Participants"


src/app.js
(function($) {
"use strict";

var participants = Echo.App.manifest("Echo.Apps.RecentParticipants");

if (Echo.App.isDefined(participants)) return;

participants.config = {
	"targetURL": undefined,
	"presentation": {
		"avatarSize": 48, // in px
		"maxParticipants": 15,
		"maxWidth": undefined,
		"maxHeight": undefined
	},
	"dependencies": {
		"StreamServer": {
			"appkey": undefined,
			"apiBaseURL": "{%= apiBaseURLs.StreamServer.basic %}/",
			"liveUpdates": {
				"transport": "websockets",
				"enabled": true,
				"websockets": {
					"URL": "{%= apiBaseURLs.StreamServer.ws %}/"
				}
			}
		}
	}
};

participants.labels = {
	"noParticipants": "No participants yet.<br>Stay tuned!"
};

participants.dependencies = [{
	"url": "{config:cdnBaseURL.sdk}/api.pack.js",
	"control": "Echo.StreamServer.API"
}, {
	"url": "//cdn.echoenabled.com/apps/echo/conversations/v2/streamserver.pack.js",
	"control": "Echo.StreamServer.Controls.FaceCollection"
}];

participants.init = function() {
	// check for "targetURL" field, without
	// this field we are unable to retrieve any data
	if (!this.config.get("targetURL")) {
		this.showMessage({
			"type": "error",
			"message": "Unable to retrieve data, target URL is not specified."
		});
		return;
	}

	// attach unique CSS class for this app instance
	// to target this instance in CSS selectors later
	var uniqueCSSClass = this.cssPrefix + this.config.get("context");
	this.set("uniqueCSSClass", uniqueCSSClass);
	this.config.get("target").addClass(uniqueCSSClass);
	this._applyDynamicCSS();

	this.render();
	this.ready();
};

participants.templates.main =
	'<div class="{class:container}">' +
		'<div class="{class:facepile}"></div>' +
	'</div>';

participants.templates.empty =
	'<div class="{class:empty}">' +
		'<span class="{class:message}">{label:noParticipants}</span>' +
	'</div>';

participants.renderers.facepile = function(element) {
	var app = this;
	var ss = this.config.get("dependencies.StreamServer");
	this.initComponent({
		"id": "Facepile",
		"component": "Echo.StreamServer.Controls.FaceCollection",
		"config": {
			"target": element,
			"query": this._assembleQuery(),
			"item": {"avatar": true, "text": false},
			"infoMessages": {"layout": "full"},
			"appkey": ss.appkey,
			"apiBaseURL": ss.apiBaseURL,
			"liveUpdates": ss.liveUpdates,
			"maxUsersCount": this.config.get("presentation.maxParticipants"),
			"ready": function() {
				// update no participants UI
				// with consistent message and icon
				if (!this.get("count.visible")) {
					element.empty().append(app.substitute({
						"template": app.templates.empty
					}));
				}
			}
		}
	});
	return element;
};

participants.methods._applyDynamicCSS = function() {
	var app = this;
	var sizeRestrictions = Echo.Utils.foldl([], ["Width", "Height"], function(key, acc) {
		var value = app.config.get("presentation.max" + key);
		if (value) {
			acc.push("max-" + key.toLowerCase() + ": " + value + "px;");
		}
	}).join(" ");
	var css = '.{self:uniqueCSSClass} .echo-streamserver-controls-face-container .echo-streamserver-controls-face-avatar, .{self:uniqueCSSClass} .echo-streamserver-controls-face-container .echo-streamserver-controls-face-avatar img { width: {config:presentation.avatarSize}px; height: {config:presentation.avatarSize}px; }' +
	'.{self:uniqueCSSClass} .echo-streamserver-controls-facecollection { ' + sizeRestrictions + ' overflow: auto; margin: 0px auto; }';
	Echo.Utils.addCSS(this.substitute({"template": css}), this.config.get("context"));
};

participants.methods._assembleQuery = function() {
	var query =
		"childrenof:{config:targetURL} " +
		"itemsPerPage:{config:presentation.maxParticipants} " +
		"sortOrder:reverseChronological " +
		"children:0";
	return this.substitute({"template": query});
};

participants.css =
	'.{class} .echo-streamserver-controls-facecollection-and,' +
		'.{class} .echo-streamserver-controls-facecollection-more { display: none; }' +
	'.{class} .echo-streamserver-controls-face-container .echo-streamserver-controls-face-avatar { margin: 5px; vertical-align: middle; }' +
	'.{class:empty} { border: 1px solid #d2d2d2; background-color: #fff; margin: 0px; margin-bottom: 10px; padding: 30px 20px; text-align: center; }' +
	'.{class:empty} .{class:message} { background: url("//cdn.echoenabled.com/apps/echo/conversations/v2/sdk-derived/images/info.png") no-repeat; margin: 0 auto; font-size: 14px; font-family: "Helvetica Neue", Helvetica, "Open Sans", sans-serif; padding-left: 40px; display: inline-block; text-align: left; width: 140px; line-height: 16px; color: #7f7f7f; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; max-width: 100%; box-sizing: border-box; }';

Echo.App.create(participants);

})(Echo.jQuery);


src/dashboard.js
(function($) {
"use strict";

if (Echo.AppServer.Dashboard.isDefined("Echo.Apps.RecentParticipants.Dashboard")) return;

var dashboard = Echo.AppServer.Dashboard.manifest("Echo.Apps.RecentParticipants.Dashboard");

dashboard.inherits = Echo.Utils.getComponent("Echo.AppServer.Dashboards.AppSettings");

dashboard.mappings = {
	"dependencies.appkey": {
		"key": "dependencies.StreamServer.appkey"
	}
};

dashboard.dependencies = [{
	"url": "{config:cdnBaseURL.apps.appserver}/controls/configurator.js",
	"control": "Echo.AppServer.Controls.Configurator"
}, {
	"url": "{config:cdnBaseURL.apps.dataserver}/dashboards.pack.js",
	"control": "Echo.DataServer.Controls.Pack"
}, {
	"url": "//cdn.echoenabled.com/apps/echo/social-map/v1/slider.js"
}];

dashboard.config.ecl = [{
	"component": "Group",
	"name": "presentation",
	"type": "object",
	"config": {
		"title": "Presentation"
	},
	"items": [{
		"component": "Slider",
		"name": "avatarSize",
		"type": "number",
		"default": 48,
		"config": {
			"title": "Avatar size",
			"desc": "Specifies user avatar size (in px)",
			"min": 10,
			"max": 100,
			"step": 1,
			"unit": "px"
		}
	}, {
		"component": "Slider",
		"name": "maxParticipants",
		"type": "number",
		"default": 15,
		"config": {
			"title": "Maximum participants",
			"desc": "Specifies maximum amount of participants to be displayed at app load time. Note: new participants will be attached to the list.",
			"min": 5,
			"max": 50,
			"step": 1
		}
	}, {
		"component": "Input",
		"name": "maxWidth",
		"type": "number",
		"config": {
			"title": "Maximum width",
			"desc": "Specify a maximum width (in pixels) of an App container",
			"data": {"sample": 500}
		}
	}, {
		"component": "Input",
		"name": "maxHeight",
		"type": "number",
		"config": {
			"title": "Maximum height",
			"desc": "Specify a maximum height (in pixels) of an App container",
			"data": {"sample": 300}
		}
	}]
}, {
	"component": "Group",
	"name": "dependencies",
	"type": "object",
	"config": {
		"title": "Dependencies",
		"expanded": false
	},
	"items": [{
		"component": "Select",
		"name": "appkey",
		"type": "string",
		"config": {
			"title": "StreamServer application key",
			"desc": "Specifies the application key for this instance",
			"options": []
		}
	}]
}, {
	"name": "targetURL",
	"component": "Echo.DataServer.Controls.Dashboard.DataSourceGroup",
	"type": "string",
	"required": true,
	"config": {
		"title": "",
		"expanded": false,
		"labels": {
			"dataserverBundleName": "Echo Recent Participants Auto-Generated Bundle for {instanceName}"
		},
		"apiBaseURLs": {
			"DataServer": "{%= apiBaseURLs.DataServer %}/"
		}
	}
}];

dashboard.modifiers = {
	"dependencies.appkey": {
		"endpoint": "customer/{self:user.getCustomerId}/appkeys",
		"processor": function() {
			return this.getAppkey.apply(this, arguments);
		}
	},
	"targetURL": {
		"endpoint": "customer/{self:user.getCustomerId}/subscriptions",
		"processor": function() {
			return this.getBundleTargetURL.apply(this, arguments);
		}
	}
};

dashboard.init = function() {
	this.parent();
};

Echo.AppServer.Dashboard.create(dashboard);

})(Echo.jQuery);
